src/backend/distributed/metadata/metadata_sync.c

```diff
@@ -2006,3 +2062,444 @@ ShouldInitiateMetadataSync(bool *lockFailure)
+						errmsg("Metadata syncing is only allowed for "
+							   "known replication models.")));
+	}
+}
+
+
+/*
+ * citus_internal_add_shard_metadata is an internal UDF to
+ * add a row to pg_dist_shard.
+ */
+Datum
+citus_internal_add_shard_metadata(PG_FUNCTION_ARGS)
+{
+	Oid relationId = PG_GETARG_OID(0);
+	int64 shardId = PG_GETARG_INT64(1);
+	char storageType = PG_GETARG_CHAR(2);
+
+	text *shardMinValue = NULL;
+	if (!PG_ARGISNULL(3))
+	{
+		shardMinValue = PG_GETARG_TEXT_P(3);
+	}
+
+	text *shardMaxValue = NULL;
+	if (!PG_ARGISNULL(4))
+	{
+		shardMaxValue = PG_GETARG_TEXT_P(4);
+	}
+
+	if (!ShouldSkipMetadataChecks())
+	{
+		/* only owner of the table (or superuser) is allowed to add the Citus metadata */
+		EnsureTableOwner(relationId);
+
+		/* this UDF is not allowed allowed for executing as a separate command */
+		EnsureCoordinatorInitiatedOperation();
+
+		/*
+		 * Even if the table owner is a malicious user and the shard metadata is
+		 * not sane, the user can only affect its own tables. Given that the
+		 * user is owner of the table, we should allow.
+		 */
+		EnsureShardMetadataIsSane(relationId, shardId, storageType, shardMinValue,
+								  shardMaxValue);
+	}
+
+	InsertShardRow(relationId, shardId, storageType, shardMinValue, shardMaxValue);
+
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * EnsureCoordinatorInitiatedOperation is a helper function which ensures that
+ * the execution is initiated by the coordinator on a worker node.
+ */
+static void
+EnsureCoordinatorInitiatedOperation(void)
+{
+	if (!IsCitusInitiatedRemoteBackend() || !MyBackendIsInDisributedTransaction() ||
+		GetLocalGroupId() == COORDINATOR_GROUP_ID)
+	{
+		ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+						errmsg("This is an internal function that only Citus "
+							   "requires to use in a distributed transaction")));
+	}
+}
+
+
+/*
+ * EnsureShardMetadataIsSane ensures that the input values are safe
+ * for inserting into pg_dist_shard metadata.
+ */
+static void
+EnsureShardMetadataIsSane(Oid relationId, int64 shardId, char storageType,
+						  text *shardMinValue, text *shardMaxValue)
+{
+	if (shardId <= INVALID_SHARD_ID)
+	{
+		ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+						errmsg("Invalid shard id: %ld", shardId)));
+	}
+
+	if (storageType != SHARD_STORAGE_TABLE)
+	{
+		ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+						errmsg("Invalid shard storage type: %c", storageType)));
+	}
+
+	char partitionMethod = PartitionMethodViaCatalog(relationId);
+	if (!(partitionMethod == DISTRIBUTE_BY_HASH ||
+		  partitionMethod == DISTRIBUTE_BY_NONE))
+	{
+		/* connection from the coordinator operating on a shard */
+		ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+						errmsg("Metadata syncing is only allowed for hash, "
+							   "reference and local tables: %c", partitionMethod)));
+	}
+
+	List *distShardTupleList = LookupDistShardTuples(relationId);
```

Comment:
Well, that's a good point. The simplest idea that I can come up with to acquire exclusive lock on the table. We already ensure that the user is owner of the table, so they can already acquire exclusive lock on the table. 

Do you see any issues with that?

The future side effect would be we cannot run adding shards in parallel. Though, I don't think that is a real issue.

Reply:
(See `LockRelationOid` function calls added for this purpose, in the higher level)

Reply:
> Do you see any issues with that?
> 
> The future side effect would be we cannot run adding shards in parallel. Though, I don't think that is a real issue.

Maybe I'm misunderstanding the the change, but this would also block all writes to the table right? `ShareUpdateExclusive` seems like a better fit, because it allows writes, but disallows schema changes.

Reply:
I don't mind acquiring `ExclusiveLock` because in a coordinator initiated metadata sync, this table is only visible to the distributed transaction that is creating the table. This is mostly to prevent regular users to manually corrupt the metadata.

(In case of coordinator initiated re-sync of metadata, the table is DROPed in the same transaction, so AccessExclusive is acquired).

I might add a comment to clarify further?

Reply:
Do we only call these functions when creating a distributed table? Not when, rebalancing, tenant isolation or shard splits?

Reply:
yes, we do use these locks on rebalancing and others as well. So, `ShareUpdateExclusive ` seems like a better one